### PR TITLE
Features/lysyk code editor

### PR DESCRIPTION
**Pull request by PeterLysyk**

### CODE REVIEW

File: src/editor/codeeditor.cpp [hunk 1]
```diff
@@ -13,13 +14,20 @@ CodeEditor::CodeEditor(QWidget *parent) : QPlainTextEdit(parent)
...
     //This signal is emitted when the text document needs an update of the specified rect.
     //If the text is scrolled, rect will cover the entire viewport area.
     //If the text is scrolled vertically, dy carries the amount of pixels the viewport was scrolled.
     connect(this, SIGNAL(updateRequest(QRect,int)), this, SLOT(updateLineNumberArea(QRect,int)));
-
+    connect(this, SIGNAL(textChanged()), this, SLOT(runLexer()));
     connect(this, SIGNAL(cursorPositionChanged()), this, SLOT(highlightCurrentLine()));
+    connect(timer, SIGNAL(timeout()), this, SLOT(saveStateInTheHistory()));
+    connect(this, SIGNAL(textChanged()), this, SLOT(changesAppeared()));
+
```
Comment (vsakharevych) [inline src/editor/codeeditor.cpp:27]: In line 26 of code you connected signal with other signal

Comment (PeterLysyk) [reply src/editor/codeeditor.cpp:27]: That was test signal. Just forgot to delete it)

File: src/editor/changesmanager.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+void ChangeManager::removeHistory()//it's not necessary to hold history if we returned to the start and didn't press ctrl + Y
+{
+    if(!changesHistory.size())
+        return;
+    auto it_victim = changesHistory.end() - 1;// move from the end of list
+    while(it_victim != currentState_it)//delete all till pass iterator position(iterator isn't point to the end only if we pressed ctrl + Z)
+    {
+        it_victim--;
+        changesHistory.pop_back();
+    }
```
Comment (vsakharevych) [inline src/editor/changesmanager.cpp]: this loop might be replaced with:
changesHistory.erase(std::next(currentState_it, 1), changesHistory.end());

File: src/editor/changesmanager.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+std::string ChangeManager::redo()
+{
+    if (currentState_it == (changesHistory.end() - 1))//if is no any record in the change history list
+        return currentFileState;
+
+    currentState_it++;
+
+    auto pos = currentState_it->begin_change_pos;
+    std::string to   = currentState_it->after;
```
Comment (vsakharevych) [inline src/editor/changesmanager.cpp:102]: it might have sense to check if currentState_it is valid before using it

File: src/editor/hotkeys.cpp [hunk 4]
```diff
@@ -2,46 +2,72 @@
...
-            setViewportMargins(lineNumberAreaWidth(), 0, 0, 0);// reset text margin in accordance to linecouter change
-            return;
-        }
+    if((e->key() == Qt::Key_Enter || e->key() == Qt::Key_Return) &&// shift + enter
+            ((e->modifiers() & Qt::ShiftModifier) == Qt::ShiftModifier))
+    {
+        e = new QKeyEvent(e->type(), e->key(), e->modifiers()&Qt::MetaModifier &Qt::KeypadModifier);
+
+        std::string str = this->toPlainText().toUtf8().constData();
+    }
```
Comment (vsakharevych) [inline src/editor/hotkeys.cpp:11]: return after this block

File: src/editor/hotkeys.cpp [hunk 5]
```diff
@@ -2,46 +2,72 @@
...
+
+    if(e->key() == Qt::Key_Space || e->key() == Qt::Key_Enter || e->key() == Qt::Key_Return)
+    {
+       saveStateInTheHistory();
+    }
```
Comment (vsakharevych) [inline src/editor/hotkeys.cpp:18]: return after this block

File: src/editor/hotkeys.cpp [hunk 6]
```diff
@@ -2,46 +2,72 @@
...
+
+
+    if((e->key() == Qt::Key_V && e->modifiers() & Qt::ControlModifier))
+    {
+        saveStateInTheHistory();
+    }
```
Comment (vsakharevych) [inline src/editor/hotkeys.cpp:26]: return after this block

**Review by vsakharevych — APPROVED** [review]
Please consider added comments before merging

File: src/editor/changesmanager.cpp (hunk 3 above)
Comment (PeterLysyk) [reply src/editor/changesmanager.cpp:102]: I didn't consider the situation where is only one element in the deque. Fixed it

File: src/editor/changemanager.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
+#ifndef CHANGESMANAGER_H
+#define CHANGESMANAGER_H
+
+#define CHANGES_HISTORY_MAX_SIZE 200
+
+#include<string>
```
Comment (opastushkov) [inline src/editor/changemanager.h:6]: Use Qt classes.

File: src/editor/changemanager.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
+#ifndef CHANGESMANAGER_H
+#define CHANGESMANAGER_H
+
+#define CHANGES_HISTORY_MAX_SIZE 200
```
Comment (opastushkov) [inline src/editor/changemanager.h]: Use const.

**Review by opastushkov — COMMENTED** [review]
Use Qt classes. Instead implement own redo/undo you can use QUndoStack, QUndoCommand  & etc.

File: src/editor/changemanager.h (hunk 7 above)
Comment (PeterLysyk) [reply src/editor/changemanager.h:6]: I prefere to use std

File: src/editor/changemanager.h (hunk 8 above)
Comment (PeterLysyk) [reply src/editor/changemanager.h]: ok

**Review by MarsLviv — APPROVED** [review]

File: src/editor/changesmanager.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+std::string ChangeManager::redo()
+{
+    if (currentState_it == (changesHistory.end() - 1) ||
+        currentState_it == (changesHistory.end()))//if is no any record in the change history list
+        return currentFileState;
+
+    currentState_it++;
+
+    auto pos = currentState_it->begin_change_pos;
```
Comment (FedorenkoAndriy) [inline src/editor/changesmanager.cpp:101]: Do you need this variable pos? You are using it only once line 105

static_cast<long long>(pos);        ->     static_cast<long long>(currentState_it->begin_change_pos);

**Review by FedorenkoAndriy — COMMENTED** [review]
See my comments to the previous review

**Pull request merged**